telegram/ext/_picklepersistence.py

```diff
@@ -26,14 +28,99 @@
     Tuple,
     overload,
     cast,
+    Type,
+    Set,
+    Callable,
 )
 
+from telegram import Bot, TelegramObject
 from telegram._utils.types import FilePathInput
+from telegram._utils.warnings import warn
 from telegram.ext import BasePersistence, PersistenceInput
 from telegram.ext._contexttypes import ContextTypes
 from telegram.ext._utils.types import UD, CD, BD, ConversationDict, CDCData
 
 
+_REPLACED_KNOWN_BOT = "a known bot replaced by PTB's PicklePersistence"
+_REPLACED_UNKNOWN_BOT = "an unknown bot replaced by PTB's PicklePersistence"
+
+
+def _all_subclasses(cls: Type[TelegramObject]) -> Set[Type[TelegramObject]]:
+    """Gets all subclasses of the specified object, recursively. from
+    https://stackoverflow.com/a/3862957/9706202
+    """
+    subclasses = cls.__subclasses__()
+    return set(subclasses).union([s for c in subclasses for s in _all_subclasses(c)])
+
+
+def _reconstruct_to(cls: Type[TelegramObject], kwargs: dict) -> TelegramObject:
+    """
+    This method is used for unpickling. The data, which is in the form a dictionary, is
+    converted back into a class. Functions the same as :meth:`TelegramObject.__setstate__`
+    """
+    obj = cls(**kwargs)
+    if '_bot' in kwargs:
+        obj.set_bot(kwargs['_bot'])
+    return obj
+
+
+def _custom_reduction(cls: TelegramObject) -> Tuple[Callable, Tuple[Type[TelegramObject], dict]]:
+    """
+    This method is used for pickling. The bot attribute is preserved so _BotPickler().persistent_id
+    works as intended.
+    """
+    # Adapted from self.to_dict(), unlike to_dict(), this is not recursive.
+    data = {}
+    attrs = {attr for cls in cls.__class__.__mro__[:-1] for attr in cls.__slots__}
+    for key in attrs:
+        value = getattr(cls, key, None)
+        if value is not None:
+            data[key] = value
```

Comment:
why not just call `__getstate__` here? IMO the minor break of encapsulation is justifiable since it removes code duplication. I'm not even sure, if calling magic methods is considered calling protected methods 🤔

Reply:
I'll unify all 3 methods as said above

> I'm not even sure if...

Don't think so. We call `__mro__ ` in there and that's standard practice